File: mobile/lib/core/unread/last_seen_store.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// Persisted "have I seen this?" state backing the bottom-nav unread dots.
/// Shared by every module (Nutrition/Training/Forms) instead of each one
/// inventing its own storage — the only two shapes any of them need are
/// "one identity key" (a plan) or "a set of ids" (form requests). Survives
/// app restart by design (backed by shared_preferences, same as the rest of
/// the app's local persistence — see AccessCache/LocaleController).
class LastSeenStore {
  Future<String?> getPlanKey(String module) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('last_seen_plan_$module');
  }

  Future<void> setPlanKey(String module, String key) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('last_seen_plan_$module', key);
  }

  Future<Set<String>> getSeenIds(String module) async {
    final prefs = await SharedPreferences.getInstance();
    return (prefs.getStringList('last_seen_ids_$module') ?? const []).toSet();
  }

  Future<void> setSeenIds(String module, Set<String> ids) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList('last_seen_ids_$module', ids.toList());
  }
}

final lastSeenStoreProvider = Provider<LastSeenStore>((ref) => LastSeenStore());

File: client/lib/lastSeenStore.js
// Persisted "have I seen this?" state backing the bottom-nav unread dots.
// Shared by every module (Nutrition/Training/Forms) instead of each one
// inventing its own storage — direct web analog of the mobile app's
// LastSeenStore (shared_preferences there, localStorage here).

const PLAN_KEY_PREFIX = "ff_last_seen_plan_";
const IDS_KEY_PREFIX  = "ff_last_seen_ids_";

// Fired whenever a module marks itself seen, so the nav can recheck badges
// immediately instead of waiting for its next poll tick.
export const NAV_BADGES_REFRESH_EVENT = "ff:nav-badges-refresh";

function notifyRefresh() {
    if (typeof window !== "undefined") window.dispatchEvent(new Event(NAV_BADGES_REFRESH_EVENT));
}

// A plan's identity for "have I seen this?" purposes: its id plus
// activated_at. Changes on activation/restart (a coach publishing a new
// plan), not on ordinary content edits to the same active plan.
export function planKey(id, activatedAt) {
    return `${id}|${activatedAt ?? ""}`;
}

export function getSeenPlanKey(moduleName) {
    try {
        return localStorage.getItem(`${PLAN_KEY_PREFIX}${moduleName}`);
    } catch {
        return null;
    }
}

export function setSeenPlanKey(moduleName, key) {
    try {
        localStorage.setItem(`${PLAN_KEY_PREFIX}${moduleName}`, key);
    } catch { /* storage unavailable — badge just won't clear this session */ }
    notifyRefresh();
}

export function getSeenIds(moduleName) {
    try {
        const raw = JSON.parse(localStorage.getItem(`${IDS_KEY_PREFIX}${moduleName}`) || "[]");
        return new Set(Array.isArray(raw) ? raw : []);
    } catch {
        return new Set();
    }
}

export function setSeenIds(moduleName, ids) {
    try {
        localStorage.setItem(`${IDS_KEY_PREFIX}${moduleName}`, JSON.stringify([...ids]));
    } catch { /* storage unavailable — badge just won't clear this session */ }
    notifyRefresh();
}
